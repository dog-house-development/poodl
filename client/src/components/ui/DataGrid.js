import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import Loading from './Loading';
import Field from './Field';
import Utils from '../../utils/Utils';

const propTypes = {
    data: PropTypes.array.isRequired,
    loading: PropTypes.bool,
    onRowClick: PropTypes.func,
    includeFilterControls: PropTypes.bool
};

const defaultProps = {
    loading: false,
    includeFilterControls: true
};

class DataGrid extends Component {
    constructor(props) {
        super(props);
        this.state = { filterValue: '', filteredData: props.data };
    }

    componentDidUpdate(prevProps) {
        if (!_.isEqual(this.props.data, prevProps.data)) {
            this.setState({ filteredData: this.props.data });
        }
    }

    getHeaderMarkup() {
        return (
            <tr>
                {_.map(_.keys(_.get(this.props, 'data[0]')), key => {
                    return key !== 'key' ? <th key={key}>{_.startCase(key)}</th> : null;
                })}
            </tr>
        );
    }

    handleFilterChange = event => {
        event.preventDefault();
        this.setState({
            filterValue: event.target.value
        });
    };

    getFilterMarkup() {
        if (this.props.includeFilterControls) {
            return (
                <Field
                    className="filter-field"
                    kind="secondary"
                    onChange={this.handleFilterChange}
                    id="filter"
                    value={this.state.filterValue}
                    placeholder="Search"
                    autoComplete="off"
                />
            );
        }
    }

    getFilteredData = () => Utils.searchCollection(this.state.filterValue, this.props.data);

    getBodyMarkup = () =>
        _.map(this.getFilteredData(), row => (
            <tr key={row.key} onClick={e => this.props.onRowClick(e, row.key)}>
                {_.map(row, (value, key) => (key !== 'key' ? <td key={key}>{value}</td> : null))}
            </tr>
        ));

    getTableMarkup() {
        if (this.props.loading) {
            return <Loading />;
        }
        return (
            <div className="panel datagrid">
                <table>
                    <thead>{this.getHeaderMarkup()}</thead>
                    <tbody>{this.getBodyMarkup()}</tbody>
                </table>
            </div>
        );
    }

    render() {
        return (
            <div>
                {this.getFilterMarkup()}
                {this.getTableMarkup()}
            </div>
        );
    }
}

DataGrid.propTypes = propTypes;
DataGrid.defaultProps = defaultProps;
export default DataGrid;
