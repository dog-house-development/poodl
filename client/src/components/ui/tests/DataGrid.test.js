import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';

import { DataGrid } from '../DataGrid';

configure({ adapter: new Adapter() });

describe('DataGrid tests', () => {
    let wrapper, instance;
    const setInstanceAndWrapper = (_props = {}) => {
        const props = _.assign(
            {},
            {
                data: [
                    { name: 'Dong Nong', email: 'dongnong@nowhere.com', key: '123' },
                    { name: 'Juck Gruck', email: 'juckgruck@nowhere.com', key: '321' }
                ],
                loading: false,
                location: { search: {} }
            },
            _props
        );
        wrapper = shallow(<DataGrid {..._.assign({}, props)} />);
        instance = wrapper.instance();
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('componentDidUpdate', () => {
        it('should set filteredData to props data if it is next', () => {
            setInstanceAndWrapper({ data: [{ name: 'Dong Nong', email: 'dongnong@nowhere.com', key: '123' }] });
            instance.componentDidUpdate({ data: { name: 'Dang Nang', email: 'dangnang@nowhere.com', key: '123' } });
            expect(instance.state.filteredData).toEqual([
                {
                    name: 'Dong Nong',
                    email: 'dongnong@nowhere.com',
                    key: '123'
                }
            ]);
        });
    });

    describe('getSortedData', () => {
        it('should set filteredData to props data if it is next', () => {
            expect(instance.state.filterValue).toEqual('');
            instance.setState({ filterValue: 'Dong' });
            instance.getSortedData();
            expect(instance.getSortedData()).toEqual([
                {
                    name: 'Dong Nong',
                    email: 'dongnong@nowhere.com',
                    key: '123'
                }
            ]);
        });
    });

    describe('handleFilterChange', () => {
        it('should handle change to filter', () => {
            expect(wrapper.find('Field').at(0)).toMatchSnapshot();
            const e = {
                target: {
                    id: 'FilterValue',
                    value: 'test'
                }
            };
            wrapper
                .find('Field')
                .first()
                .simulate('change', e);
            expect(instance.state.filterValue).toEqual('test');
        });
    });
    describe('getFilterMarkup', () => {
        it('should return filter field', () => {
            setInstanceAndWrapper({ includeFilterControls: true });
            instance.getFilterMarkup();
            expect(instance.getFilterMarkup()).toMatchSnapshot();
        });
    });

    describe('render', () => {
        it('should render loading', () => {
            setInstanceAndWrapper({ loading: true });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render data', () => {
            setInstanceAndWrapper();
            expect(wrapper).toMatchSnapshot();
        });
    });
});
