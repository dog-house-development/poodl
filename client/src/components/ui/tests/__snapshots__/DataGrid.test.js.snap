// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataGrid tests getFilterMarkup should return filter field 1`] = `
<Field
  autoComplete="off"
  className="filter-field"
  clearable={true}
  id="filter"
  kind="secondary"
  leftIcon="search"
  onChange={[Function]}
  placeholder="Search"
  size="normal"
  type="text"
  value=""
/>
`;

exports[`DataGrid tests handleFilterChange should handle change to filter 1`] = `
<Field
  autoComplete="off"
  className="filter-field"
  clearable={true}
  id="filter"
  kind="secondary"
  leftIcon="search"
  onChange={[Function]}
  placeholder="Search"
  size="normal"
  type="text"
  value=""
/>
`;

exports[`DataGrid tests render should render data 1`] = `
<div>
  <Field
    autoComplete="off"
    className="filter-field"
    clearable={true}
    id="filter"
    kind="secondary"
    leftIcon="search"
    onChange={[Function]}
    placeholder="Search"
    size="normal"
    type="text"
    value=""
  />
  <div
    className="panel datagrid"
  >
    <table>
      <thead>
        <tr>
          <th
            key="name"
          >
            Name
          </th>
          <th
            key="email"
          >
            Email
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="no-click"
          key="123"
          onClick={[Function]}
        >
          <td
            key="name"
          >
            Dong Nong
          </td>
          <td
            key="email"
          >
            dongnong@nowhere.com
          </td>
        </tr>
        <tr
          className="no-click"
          key="321"
          onClick={[Function]}
        >
          <td
            key="name"
          >
            Juck Gruck
          </td>
          <td
            key="email"
          >
            juckgruck@nowhere.com
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`DataGrid tests render should render loading 1`] = `
<div>
  <Field
    autoComplete="off"
    className="filter-field"
    clearable={true}
    id="filter"
    kind="secondary"
    leftIcon="search"
    onChange={[Function]}
    placeholder="Search"
    size="normal"
    type="text"
    value=""
  />
  <Loading
    content="Loading..."
    kind="primary"
    size="medium"
  />
</div>
`;
